kernel/sched.c

```diff
@@ -994,9 +1025,12 @@ void z_impl_k_thread_priority_set(k_tid_t thread, int prio)
 
 	bool need_sched = z_thread_prio_set((struct k_thread *)thread, prio);
 
-	flag_ipi();
-	if (need_sched && _current->base.sched_locked == 0U) {
-		z_reschedule_unlocked();
+	if (need_sched) {
+		if (_current->base.sched_locked == 0U) {
+			z_reschedule_unlocked();
+		} else {
+			signal_pending_ipi();
+		}
```

Comment:
Had to dig here trying to figure out what was going on.  IPI logic absolutely doesn't belong up here in k_thread_priority_set().   I think I get what you're trying to do: we had a gate on the reschedule here based on "sched_locked"[1].  But in SMP that's a little wrong, because even if you can't reschedule locally you still need to tell other CPUs that the local CPU has reached a scheduling point.

But the fix there is to remove the sched_locked logic in SMP[2], or arguably entirely (k_sched_lock() is a very rarely used API, I'd expect it's not worth the extra test here).

Basically what you're actually doing here is working around a pre-existing bug.  Let's fix that instead.

[1] Which is a pretty subtle thing.  It's a holdover from way back in the Ben Walsh era that reflects the top bits of the 16 bit priority such that you can decrement it to make it negative and thus create a "maximally low" priority value.

[2] The schedule lock simply doesn't work in SMP absent great care, and in any case this is just a performance optimization.  All the normal priority logic will prevent preemption of the current thread, we don't actually need this code here at all.

Reply:
If I follow both you and the code correctly, the `(_current->base.sched_locked == 0U)` in k_thread_priority_set() exists only for UP platforms that use arch_swap(). Aside--the arch_switch code is both so much cleaner and easier to follow.

Digging in a little more suggests do_swap() is the correct place to signal the IPI. However, the present version of do_swap() is lacking as it only signals any flagged IPIs if a new thread is to be switched in (see `z_requeue_current()`). But since the current thread could be schedule locked, if there any IPIs flagged for signalling even if the current thread is not going to be switched out, then these should still be signalled. For what it is worth, `z_get_next_switch_handle()` does not have this issue.

Updated code coming soon.

Reply:
Oops ... I wrote too soon. I forgot about `z_swap_next_thread()`. It already does the `signal_pending_ipi()`.